# Accounts Receivable (AR) & Payment Collection Analysis

## Data Validation & Cleaning

### Objective

The objective of this notebook is to validate the quality of the Accounts Receivable datasets before analysis.

The validation process includes:

- Importing datasets
- Inspecting data structure
- Identifying missing values
- Detecting duplicate records
- Removing unnecessary columns
- Validating relationships between datasets

The cleaned datasets will be used for SQL analysis, Python EDA, and Power BI dashboard development.

In [42]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)

In [43]:
customers = pd.read_csv("Customer.csv")
invoices = pd.read_csv("invoice.csv")
payments = pd.read_csv("Payment.csv")
sales = pd.read_csv("Sales.csv")
credit = pd.read_csv("Credits.csv")

In [44]:
datasets = {
    "Customers": customers,
    "Invoices": invoices,
    "Payments": payments,
    "Sales": sales,
    "Credit Limits": credit
}

for name, df in datasets.items():
    print(f"{name}: {df.shape}")

Customers: (500, 8)
Invoices: (8005, 6)
Payments: (6748, 6)
Sales: (10000, 5)
Credit Limits: (500, 5)


In [45]:
customers.head()

,Customer_ID,Customer_Name,Region,Industry,Credit_Limit,Payment_Terms,Account_Manager,Unnamed: 7
0,C0001,Customer 1,North,Retail,438786,Net 30,Manager 4,NaN
1,C0002,Customer 2,South,Manufacturing,436123,Net 15,Manager 9,NaN
2,C0003,Customer 3,North,Wholesale,271208,Net 15,Manager 1,NaN
3,C0004,Customer 4,North,Manufacturing,171981,Net 60,Manager 10,NaN
4,C0005,Customer 5,North,Wholesale,154248,Net 60,Manager 9,NaN


In [46]:
invoices.head()

,Invoice_ID,Customer_ID,Invoice_Date,Due_Date,Invoice_Amount,Status
0,INV00001,C0383,2025-02-05,2025-03-07,42847.0,Paid
1,INV00002,C0198,2024-04-25,2024-05-10,33240.0,Paid
2,INV00003,C0185,2024-11-20,2024-12-05,25851.0,Outstanding
3,INV00004,C0292,2025-03-06,2025-04-05,36306.0,Paid
4,INV00005,C0009,2025-08-24,2025-09-08,1403.0,Paid


In [47]:
invoices = invoices.dropna(subset=["Invoice_Amount"])

invoices.drop_duplicates(inplace=True)

In [48]:
invoices.isnull().sum()[invoices.isnull().sum() > 0]

Series([], dtype: int64)

In [49]:
invoices[invoices.duplicated(subset=["Customer_ID"], keep=False)]

,Invoice_ID,Customer_ID,Invoice_Date,Due_Date,Invoice_Amount,Status
0,INV00001,C0383,2025-02-05,2025-03-07,42847.0,Paid
1,INV00002,C0198,2024-04-25,2024-05-10,33240.0,Paid
2,INV00003,C0185,2024-11-20,2024-12-05,25851.0,Outstanding
3,INV00004,C0292,2025-03-06,2025-04-05,36306.0,Paid
4,INV00005,C0009,2025-08-24,2025-09-08,1403.0,Paid
...,...,...,...,...,...,...
7995,INV07996,C0370,2024-06-16,2024-07-01,22641.0,Paid
7996,INV07997,C0075,2025-05-11,2025-07-10,2296.0,Outstanding
7997,INV07998,C0442,2025-12-23,2026-01-22,3276.0,Outstanding
7998,INV07999,C0352,2024-04-08,2024-05-08,20114.0,Paid


In [50]:
payments.head()
sales.head()
credit.head()

,Customer_ID,Credit_Limit,Current_Exposure,Risk_Rating,Utilization
0,C0001,438786,247729,Low,56.46%
1,C0002,436123,279089,Medium,63.99%
2,C0003,271208,249493,High,91.99%
3,C0004,171981,25135,High,14.61%
4,C0005,154248,131209,High,85.06%


In [51]:
for name, df in datasets.items():
    print(f"\n{name}")
    print(df.isnull().sum())


Customers
Customer_ID          0
Customer_Name        0
Region               0
Industry             0
Credit_Limit         0
Payment_Terms        0
Account_Manager      0
Unnamed: 7         499
dtype: int64

Invoices
Invoice_ID        0
Customer_ID       0
Invoice_Date      0
Due_Date          0
Invoice_Amount    6
Status            0
dtype: int64

Payments
Payment_ID        0
Invoice_ID        0
Payment_Date      0
Month             0
Amount_Paid       0
Payment_Method    0
dtype: int64

Sales
Order_ID            0
Customer_ID         0
Sales_Date          0
Sales_Amount        0
Product_Category    0
dtype: int64

Credit Limits
Customer_ID         0
Credit_Limit        0
Current_Exposure    0
Risk_Rating         0
Utilization         0
dtype: int64


In [52]:
customers = customers.loc[:, ~customers.columns.str.contains("^Unnamed")]

In [53]:
customers.columns

Index(['Customer_ID', 'Customer_Name', 'Region', 'Industry', 'Credit_Limit',
       'Payment_Terms', 'Account_Manager'],
      dtype='str')

In [54]:
invoices[invoices["Invoice_Amount"].isnull()]

,Invoice_ID,Customer_ID,Invoice_Date,Due_Date,Invoice_Amount,Status


In [55]:
print("Customers:", customers.duplicated().sum())
print("Invoices:", invoices.duplicated().sum())
print("Payments:", payments.duplicated().sum())
print("Sales:", sales.duplicated().sum())
print("Credit:", credit.duplicated().sum())

Customers: 0
Invoices: 0
Payments: 0
Sales: 0
Credit: 0


In [56]:
for name, df in datasets.items():
    print(f"\n{name}")
    print(df.dtypes)


Customers
Customer_ID            str
Customer_Name          str
Region                 str
Industry               str
Credit_Limit         int64
Payment_Terms          str
Account_Manager        str
Unnamed: 7         float64
dtype: object

Invoices
Invoice_ID            str
Customer_ID           str
Invoice_Date          str
Due_Date              str
Invoice_Amount    float64
Status                str
dtype: object

Payments
Payment_ID          str
Invoice_ID          str
Payment_Date        str
Month               str
Amount_Paid       int64
Payment_Method      str
dtype: object

Sales
Order_ID              str
Customer_ID           str
Sales_Date            str
Sales_Amount        int64
Product_Category      str
dtype: object

Credit Limits
Customer_ID           str
Credit_Limit        int64
Current_Exposure    int64
Risk_Rating           str
Utilization           str
dtype: object


In [57]:
customers = customers.loc[:, ~customers.columns.str.contains("^Unnamed")]

In [59]:
invoices["Invoice_Date"] = pd.to_datetime(
    invoices["Invoice_Date"],
    format="mixed",
    dayfirst=True,
    errors="coerce"
)

invoices["Due_Date"] = pd.to_datetime(
    invoices["Due_Date"],
    format="mixed",
    dayfirst=True,
    errors="coerce"
)

payments["Payment_Date"] = pd.to_datetime(
    payments["Payment_Date"],
    format="mixed",
    dayfirst=True,
    errors="coerce"
)

sales["Sales_Date"] = pd.to_datetime(
    sales["Sales_Date"],
    format="mixed",
    dayfirst=True,
    errors="coerce"
)

In [60]:
print(invoices.isnull().sum())
print(payments.isnull().sum())
print(sales.isnull().sum())

Invoice_ID        0
Customer_ID       0
Invoice_Date      0
Due_Date          0
Invoice_Amount    0
Status            0
dtype: int64
Payment_ID        0
Invoice_ID        0
Payment_Date      0
Month             0
Amount_Paid       0
Payment_Method    0
dtype: int64
Order_ID            0
Customer_ID         0
Sales_Date          0
Sales_Amount        0
Product_Category    0
dtype: int64


In [61]:
missing_customers = invoices[
    ~invoices["Customer_ID"].isin(customers["Customer_ID"])
]

print("Invoices without matching customer:", len(missing_customers))

missing_customers.head()

Invoices without matching customer: 0


,Invoice_ID,Customer_ID,Invoice_Date,Due_Date,Invoice_Amount,Status


In [62]:
missing_payments = invoices[
    ~invoices["Invoice_ID"].isin(payments["Invoice_ID"])
]

print("Invoices without payment:", len(missing_payments))

missing_payments.head()

Invoices without payment: 1249


,Invoice_ID,Customer_ID,Invoice_Date,Due_Date,Invoice_Amount,Status
2,INV00003,C0185,2024-11-20,2024-05-12,25851.0,Outstanding
16,INV00017,C0189,2024-10-06,2024-10-07,7243.0,Outstanding
22,INV00023,C0489,2025-02-14,2025-01-03,30980.0,Outstanding
36,INV00037,C0047,2024-10-06,2024-09-08,20437.0,Outstanding
37,INV00038,C0357,2025-09-09,2025-09-24,14591.0,Outstanding


In [63]:
missing_sales_customer = sales[
    ~sales["Customer_ID"].isin(customers["Customer_ID"])
]

print("Sales with invalid customer:", len(missing_sales_customer))

missing_sales_customer.head()

Sales with invalid customer: 0


,Order_ID,Customer_ID,Sales_Date,Sales_Amount,Product_Category


In [64]:
missing_credit_customer = credit[
    ~credit["Customer_ID"].isin(customers["Customer_ID"])
]

print("Credit records without customer:", len(missing_credit_customer))

missing_credit_customer.head()

Credit records without customer: 0


,Customer_ID,Credit_Limit,Current_Exposure,Risk_Rating,Utilization


In [65]:
print("=" * 50)
print("RELATIONSHIP VALIDATION SUMMARY")
print("=" * 50)

print(f"Invoice → Customer : {len(missing_customers)} unmatched")
print(f"Invoice → Payment  : {len(missing_payments)} unpaid invoices")
print(f"Sales → Customer   : {len(missing_sales_customer)} unmatched")
print(f"Credit → Customer  : {len(missing_credit_customer)} unmatched")

RELATIONSHIP VALIDATION SUMMARY
Invoice → Customer : 0 unmatched
Invoice → Payment  : 1249 unpaid invoices
Sales → Customer   : 0 unmatched
Credit → Customer  : 0 unmatched


In [66]:
missing_invoices = payments[
    ~payments["Invoice_ID"].isin(invoices["Invoice_ID"])
]

print("Payments without matching invoice:", len(missing_invoices))
missing_invoices.head()

Payments without matching invoice: 3


,Payment_ID,Invoice_ID,Payment_Date,Month,Amount_Paid,Payment_Method
46,PAY00047,INV00053,2025-07-10,Oct,21576,Bank Transfer
3350,PAY03351,INV03959,2024-10-05,May,42256,Bank Transfer
6443,PAY06444,INV07629,2025-08-29,Aug,2528,UPI


In [67]:
missing_invoices

,Payment_ID,Invoice_ID,Payment_Date,Month,Amount_Paid,Payment_Method
46,PAY00047,INV00053,2025-07-10,Oct,21576,Bank Transfer
3350,PAY03351,INV03959,2024-10-05,May,42256,Bank Transfer
6443,PAY06444,INV07629,2025-08-29,Aug,2528,UPI


## Relationship Validation Findings

Three payment records did not have matching invoice records after data cleaning.

Investigation showed that the corresponding invoices were removed because they contained missing invoice amounts or duplicate records.

These payment records were retained in the original payment dataset for audit purposes but excluded from invoice-based analysis to maintain data integrity.

In [68]:
payments = payments[
    payments["Invoice_ID"].isin(invoices["Invoice_ID"])
]

In [70]:
payments["Payment_ID"].duplicated().sum()

np.int64(0)

In [71]:
payments[payments["Amount_Paid"] <= 0]

,Payment_ID,Invoice_ID,Payment_Date,Month,Amount_Paid,Payment_Method


In [72]:
payments.describe()

,Payment_Date,Amount_Paid
count,6745,6745.000000
mean,2025-02-26 07:31:44.877687,24071.749148
min,2024-01-02 00:00:00,210.000000
25%,2024-08-15 00:00:00,11693.000000
50%,2025-03-01 00:00:00,23561.000000
75%,2025-08-31 00:00:00,36069.000000
max,2026-12-03 00:00:00,49980.000000
std,NaN,14198.595085


In [73]:
customers.to_csv("customers_clean.csv", index=False)
invoices.to_csv("invoices_clean.csv", index=False)
payments.to_csv("payments_clean.csv", index=False)
sales.to_csv("sales_clean.csv", index=False)
credit.to_csv("credit_clean.csv", index=False)

print("All cleaned datasets saved successfully!")

All cleaned datasets saved successfully!


# Notebook 1: Data Validation & Cleaning Report
## Project Overview
### Objective

This notebook focuses on validating and preparing the datasets for the Accounts Receivable (AR) & Payment Collection Analysis project.

The objective is to ensure that the data is complete, accurate, and suitable for financial analysis by identifying data quality issues, validating relationships between datasets, and creating analysis-ready files.

The cleaned datasets produced in this notebook will be used for SQL analysis, Python exploratory data analysis (EDA), and Power BI dashboard development.

## Data Quality Assessment

### The following validation checks were performed before analysis:

- Imported all datasets into Python using Pandas.
- Reviewed dataset dimensions and structure.
- Examined data types for all variables.
- Identified missing values.
- Detected duplicate records.
- Removed unnecessary columns.
- Validated relationships between datasets.
- Verified referential integrity.
- Generated summary statistics to identify unusual values.

## Missing Value Treatment

### The missing value assessment identified the following issues:

- The Customer Master dataset contained one unnecessary blank column (Unnamed: 7) generated during the Excel export process. This column was removed as it contained no useful information.
- The Invoice dataset contained six records with missing Invoice_Amount values. Since these values could not be recovered from any related dataset and invoice amount is essential for financial analysis, these records were removed.

After cleaning, all remaining datasets contained no missing values in business-critical fields.

## Duplicate Record Treatment

Duplicate records were evaluated across all datasets.

The validation identified duplicate rows within the Invoice dataset.

These duplicate records were removed to ensure that each invoice represents a unique business transaction.

No duplicate records were found in the remaining datasets.

## Data Type Validation

The data types of each dataset were reviewed and standardized.

### The following conversions were performed:

- Invoice dates were converted to datetime format.
- Due dates were converted to datetime format.
- Payment dates were converted to datetime format.
- Sales dates were converted to datetime format.

Identifier columns such as Customer_ID, Invoice_ID, and Payment_ID were retained as string values to preserve data integrity.

Monetary fields remained numeric to support aggregation and financial calculations.

## Relationship Validation

Referential integrity was validated across all datasets to ensure that related records were accurately linked and suitable for analysis.

The following relationships were verified:

| Relationship | Validation Result |
|--------------|-------------------|
| Customer → Invoice | ✔️ Valid |
| Customer → Sales | ✔️ Valid |
| Customer → Credit | ✔️ Valid |
| Invoice → Payment | ✔️ Valid after data cleaning |

During the validation process, three payment records were found without corresponding invoice records. Further investigation confirmed that these invoices had been removed during the data cleaning stage due to missing `Invoice_Amount` values or duplicate records.

To maintain referential integrity and ensure consistency between the invoice and payment datasets, the three unmatched payment records were excluded from the analysis-ready dataset.

After completing these validation steps, all remaining relationships between the datasets were consistent and reliable for subsequent SQL analysis, exploratory data analysis (EDA), and Power BI reporting.

## Summary Statistics

Summary statistics were generated to verify numerical fields and identify potential anomalies.

Key observations include:

- Payment amounts ranged from 210 to 49,980.
- The average payment amount was approximately 24,072.
- Payment transactions span from January 2024 to December 2026, providing sufficient historical data for trend analysis.
- No abnormal negative payment values or unrealistic financial amounts were identified.

Overall, the financial data appears suitable for analytical modeling and reporting.

## Data Preparation Outcome

After completing the validation process:

- Missing values were resolved.
- Duplicate records were removed.
- Invalid relationships were corrected.
- Date fields were standardized.
- Clean analysis-ready datasets were exported for downstream analysis.

These datasets are now suitable for:
- Exploratory Data Analysis (EDA)
- Financial KPI calculation
- Accounts Receivable analysis
-Power BI dashboard development


## Conclusion

The data validation and cleaning process successfully prepared the Accounts Receivable datasets for analysis.

Data quality issues were identified, investigated, and resolved while maintaining the integrity of the financial records. Relationship validation confirmed that the datasets are consistent and reliable for business reporting.

The cleaned datasets provide a solid foundation for answering key business questions related to payment collection efficiency, overdue invoices, customer credit exposure, and cash flow performance.